Firebase Push Notifications.md

```diff
@@ -0,0 +1,199 @@
+# Firebase push notifications
+
+We used to send push notifications directly to APNS or GCM from our code. We were maintaining device_ids, OS types, GCM/APNS certificates and sometimes this was really tricky.
+
+We knew that an easier solution exist and then we discovered the Firebase. Firebase is a service which does all hard work for us. There is no certificates at server side and we don't need to maintain OS types, because the Firebase knows which notification goes to the APNS and which to the GCM. Basically, Firebase handles the routing and delivery of push notifications to targeted devices. Firebase also has a dashboard with the statistics and mobile developers can use this dashboard for testing purposes.
+
+There are many services which provide us with push notifications, but we are using Firebase because of its simplicity and reliability.
+
+## General
+In this chapter we will explain a few approaches for push notifications implementations and theirs PROS and CONS.
+
+It is possible to send push notification to only one device, group of devices or to topics.
+
+Before development, read chapters about [Firebase Cloud Messaging](https://firebase.google.com/docs/cloud-messaging/), and [app server development](https://firebase.google.com/docs/cloud-messaging/server).
+
+## Development Approaches
+There are three approaches for using Firebase:
+
+  * storing one device id per user
+  * storing multiple device ids per user
+  * topics
+    * With persisted topics
+    * With non persisted topics
+
+
+1. Approach based on one device id per user
+ * Using this approach means adding a new column `device_id` directly to a User model.
+ * On each login this column will be overwritten.
+ * On each logout this column should be deleted.
+
+ * PROS
+    * Easy to build.
+
+ * CONS
+    * User can get notification only on last logged in device.
```

Comment:
Mobile developers need to be reminded to do a proper logout API call when user logs out. They are known to forget to do this.